### ISSUE TITLE
Inc classification 

### ISSUE THREAD

**Issue by MiniKara-zor-el** [MiniKara-zor-el/JavaDSAlgo#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.multiclass import OneVsRestClassifier
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
import re
import pickle
import warnings
warnings.filterwarnings('ignore')

# Download necessary NLTK resources
nltk.download('stopwords')
nltk.download('wordnet')

# Step 1: Load and explore the data
def load_data(file_path):
    """Load incident data from CSV file."""
    df = pd.read_csv(file_path)
    print(f"Dataset shape: {df.shape}")
    print(f"Columns: {df.columns.tolist()}")
    return df

# Step 2: Data preprocessing and cleaning
def preprocess_data(df):
    """Clean and preprocess the incident data."""
    # Check for missing values
    print(f"Missing values:\n{df.isnull().sum()}")
    
    # Drop rows with missing summary or resolution
    df = df.dropna(subset=['summary', 'resolution'])
    
    # Combine summary and resolution for text classification
    df['combined_text'] = df['summary'] + ' ' + df['resolution']
    
    # Explore class distribution
    plt.figure(figsize=(12, 6))
    df['incident_category'].value_counts().plot(kind='bar')
    plt.title('Distribution of Incident Categories')
    plt.tight_layout()
    plt.show()
    
    return df

# Step 3: Text preprocessing
def clean_text(text):
    """Clean and normalize text data."""
    if isinstance(text, str):
        # Convert to lowercase
        text = text.lower()
        
        # Remove special characters and numbers
        text = re.sub(r'[^a-zA-Z\s]', '', text)
        
        # Tokenize
        tokens = text.split()
        
        # Remove stopwords
        stop_words = set(stopwords.words('english'))
        tokens = [word for word in tokens if word not in stop_words]
        
        # Lemmatization
        lemmatizer = WordNetLemmatizer()
        tokens = [lemmatizer.lemmatize(word) for word in tokens]
        
        # Join tokens back into text
        cleaned_text = ' '.join(tokens)
        return cleaned_text
    return ''

# Step 4: Feature engineering and model training
def build_and_train_model(df):
    """Build ML pipeline and train the model."""
    # Apply text cleaning
    df['cleaned_text'] = df['combined_text'].apply(clean_text)
    
    # Encode target variable
    label_encoder = LabelEncoder()
    y = label_encoder.fit_transform(df['incident_category'])
    
    # Save label encoder for later use
    with open('label_encoder.pkl', 'wb') as file:
        pickle.dump(label_encoder, file)
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        df['cleaned_text'], y, test_size=0.2, random_state=42, stratify=y
    )
    
    # Create a pipeline with TF-IDF and classifier
    pipeline = Pipeline([
        ('tfidf', TfidfVectorizer(max_features=5000, ngram_range=(1, 2))),
        ('classifier', OneVsRestClassifier(LogisticRegression(max_iter=1000)))
    ])
    
    # Define hyperparameters for grid search
    param_grid = {
        'tfidf__max_features': [3000, 5000],
        'classifier__estimator': [
            LogisticRegression(max_iter=1000),
            RandomForestClassifier(n_estimators=100, random_state=42)
        ]
    }
    
    # Perform grid search
    grid_search = GridSearchCV(pipeline, param_grid, cv=5, scoring='f1_weighted', n_jobs=-1)
    grid_search.fit(X_train, y_train)
    
    # Get best model
    best_model = grid_search.best_estimator_
    print(f"Best parameters: {grid_search.best_params_}")
    
    # Evaluate on test set
    y_pred = best_model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Test Accuracy: {accuracy:.4f}")
    
    # Detailed classification report
    class_names = label_encoder.classes_
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred, target_names=class_names))
    
    # Confusion matrix
    plt.figure(figsize=(10, 8))
    cm = confusion_matrix(y_test, y_pred)
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', xticklabels=class_names, yticklabels=class_names)
    plt.title('Confusion Matrix')
    plt.ylabel('True Label')
    plt.xlabel('Predicted Label')
    plt.tight_layout()
    plt.show()
    
    # Save the model
    with open('incident_classifier.pkl', 'wb') as file:
        pickle.dump(best_model, file)
    
    return best_model, label_encoder

# Step 5: Model deployment (as a prediction function)
def predict_incident_category(summary, resolution, model, label_encoder):
    """Predict incident category from summary and resolution."""
    # Combine text
    combined_text = summary + ' ' + resolution
    
    # Clean text
    cleaned_text = clean_text(combined_text)
    
    # Make prediction
    prediction_idx = model.predict([cleaned_text])[0]
    predicted_category = label_encoder.inverse_transform([prediction_idx])[0]
    
    # Get prediction probabilities
    proba = model.predict_proba([cleaned_text])[0]
    confidence = proba[prediction_idx]
    
    return {
        'predicted_category': predicted_category,
        'confidence': confidence,
        'all_probabilities': dict(zip(label_encoder.classes_, proba))
    }

# Step 6: Real-time classification API (simplified example)
def create_classification_api():
    """
    Example function showing how to set up a Flask API for incident classification.
    This is a template and needs to be expanded in a real application.
    """
    from flask import Flask, request, jsonify
    
    app = Flask(__name__)
    
    # Load model and label encoder
    with open('incident_classifier.pkl', 'rb') as file:
        model = pickle.load(file)
    
    with open('label_encoder.pkl', 'rb') as file:
        label_encoder = pickle.load(file)
    
    @app.route('/classify', methods=['POST'])
    def classify_incident():
        data = request.json
        summary = data.get('summary', '')
        resolution = data.get('resolution', '')
        
        result = predict_incident_category(summary, resolution, model, label_encoder)
        return jsonify(result)
    
    return app

# Step 7: Main function to run the entire pipeline
def main(file_path):
    """Run the complete incident classification pipeline."""
    # Load data
    df = load_data(file_path)
    
    # Preprocess data
    df = preprocess_data(df)
    
    # Build and train model
    model, label_encoder = build_and_train_model(df)
    
    # Example prediction
    example_summary = "Server is not responding to ping requests"
    example_resolution = "Restarted the server and updated network configurations"
    prediction = predict_incident_category(example_summary, example_resolution, model, label_encoder)
    
    print("\nExample Prediction:")
    print(f"Summary: {example_summary}")
    print(f"Resolution: {example_resolution}")
    print(f"Predicted Category: {prediction['predicted_category']}")
    print(f"Confidence: {prediction['confidence']:.4f}")
    
    print("\nModel and encoder saved as 'incident_classifier.pkl' and 'label_encoder.pkl'")

# Example usage
if __name__ == "__main__":
    # Replace with your actual data file path
    file_path = "incidents_data.csv"
    main(file_path)
